### ISSUE TITLE
Should -o X-mount.idmap with just a user mapping use an identity group mapping?

### ISSUE THREAD

**Issue by arachsys** [util-linux/util-linux#2175]
I've been testing the id-mapping support in util-linux libmount prior to the upcoming release.

If I bind mount a directory and specify a user mapping but not a group mapping, this creates a user namespace with no groups mapped to use for the mount mapping:

    # ls -ln /bin/ls
    -rwxr-xr-x 2 0 0 164648 Mar 20 17:49 /bin/ls
    # mount -B -o X-mount.idmap=u:0:1:100 /bin /tmp/foo
    # ls -ln /tmp/foo/ls
    -rwxr-xr-x 2 1 65534 164648 Mar 20 17:49 /tmp/foo/ls

Naively, I'd expect the groups to default to an identity mapping[1] rather than a null mapping when one isn't specified. Is this behaviour intentional?

(Presumably the behaviour is identical if a group mapping is specified without a user mapping, but I haven't checked.)

[1] i.e. map every range in /proc/self/gid_map onto itself

**Comment by arachsys** [comment]
And if it is intentional, should it be compulsory to specify a 'g' mapping if a 'u' mapping is given, and vice versa?

**Comment by arachsys** [comment]
Damn, got bitten by the horrible web interface and closed by accident. What I was trying to post was: if this is indeed intentional, should a 'g' mapping be compulsory when a 'u' mapping is given and vice versa, to avoid user confusion with the empty mappings?

**Comment by arachsys** [comment]
Another random oddity that may be unintentional: the arguments to X-mount.idmap are space-separated tokens, but an empty argument behaves differently to whitespace with no tokens.

    # mount -B -o 'X-mount.idmap= ' /bin /tmp/foo
    # ls -ln /tmp/foo/ls
    -rwxr-xr-x 2 65534 65534 164648 Mar 20 17:49 /tmp/foo/ls

vs

    # mount -B -o 'X-mount.idmap=' /bin /tmp/foo
    # ls -ln /tmp/foo/ls
   -rwxr-xr-x 2 0 0 164648 Mar 20 17:49 /tmp/foo/ls

**Comment by brauner** [comment]
Yes, it was a conscious decision to not map any ids implicitly.
I think we could provide a shortcut such as `X-mount.idmap=g:identity` though.
I'm not set on `identity`. We can also use something else. Maybe @karelzak or @poettering have an idea?

**Comment by arachsys** [comment]
Ah, okay: no problem if the default empty set is the intended behaviour here.

Do you then think for consistency the empty X-mount.idmap= option should map to empty sets of both uids and gids, in the way that 'X-mount.idmap= ' one already does - perhaps unintentionally?

Agreed, I'm not sure whether 'identity' is the right word either - I struggled a bit to come up with a really obvious one!

We have a mount(8) convenience option --mkdir (and -m) mapping to -o X-mount.mkdir, as well as the long standing --bind and --rbind. How would you and @karelzak feel about me cooking up an rfc patch for something similar for easier id-mapping? We could have something like --map-users=RANGE and --map-groups=RANGE that can each be used multiple times and compose to form a single equivalent X-mount.idmap option.

**Comment by poettering** [comment]
i so far always called the identity mapping identity mapping... term sounds fine be me.

other words could be "null mapping", "self mapping", "same mapping", but frankly i find "identity mapping" still the best.

**Comment by karelzak** [comment]
The simplification on the mount(8) command line sounds good; it will make mount(8) more compatible with unshare where we already have `--map-users` and `--map-groups`, but I'd like to have all this covered in sys-utils/mount.c only. It means converting the options to X-mount.idmap=  string and do not modify libmount.

I have no objection against `X-mount.idmap=g:identity`.

**Comment by arachsys** [comment]
One other observation: it would be nice if the shorthand syntax also allowed the equivalent of
X-mount.idmap=/proc/PID/ns/user as that's something very natural to want in 'ad hoc container' scripts. Maybe --map-users=/proc/PID/ns/user?

**Comment by brauner** [comment]
> Hmm... some question comes to mind:
> 
> * would it make sense to also support /etc/subuid and /etc/subgid for mount(8)? It's already supported in unshare(8) by --map-groups=auto
> * why in unshare(8) we do not allow to specify username/groupname when searching in /etc/sub{uid,gid}? The "auto" reads the entry for the current user only.
> 
> What about: --map-groups=@groupName or X-mount.idmap=g:@groupName to read "groupname" entry from /etc/subgid?

I personally find `/etc/subuid` and `/etc/subgid` to be broken delegation mechanisms.
I'm not sure how well it would fit with `mount(8)`. I think `--map-groups` and `--map-user` make sense but without `/etc/sub{g,u}id` integration. But just my opinion.

**Comment by karelzak** [comment]
> I personally find `/etc/subuid` and `/etc/subgid` to be broken delegation mechanisms.

OK, no problem. Fewer features, happy maintainer :-) Let's wait if anyone with a real use case will ask for this in the future.

**Comment by arachsys** [comment]
I've done some more testing of X-mount.idmap while writing the --map-user and --map-group options for mount.c.

When the mapping is invalid, hook_prepare_options() in hook_idmap.c tries to return errors. For example, something like `mount -B -o X-mount.idmap=u:0:invalid /a /b` or `mount -B -o X-mount.idmap='u:0:1:1 invalid'` would try to abort with `return -MNT_ERR_MOUNTOPT`.

However, these errors from the hook appear to be silently ignored, with the effect that a typo in an idmap silently results in an identity mapping without any indication that something potentially risky has happened.

Similarly, X-mount.mkdir=nonsense silently succeeds even though it attempts to return `-MNT_ERR_MOUNTOPT` when the strtol fails or has trailing junk.

It looks like all errors from prep-options hooks are being dropped, although I haven't dug into why that's happening yet.

**Comment by arachsys** [comment]
Spotted it:

```diff
diff --git a/libmount/src/context_mount.c b/libmount/src/context_mount.c
index 70c5043..23676cb 100644
--- a/libmount/src/context_mount.c
+++ b/libmount/src/context_mount.c
@@ -135,7 +135,7 @@ static int fix_optstr(struct libmnt_context *cxt)
                }
        }
 #endif
-       mnt_context_call_hooks(cxt, MNT_STAGE_PREP_OPTIONS);
+       rc = mnt_context_call_hooks(cxt, MNT_STAGE_PREP_OPTIONS);
 done:
        DBG(CXT, ul_debugobj(cxt, "<-- preparing options done [rc=%d]", rc));
        cxt->flags |= MNT_FL_MOUNTOPTS_FIXED;
```

I'll send a proper signed-off patch.

**Comment by arachsys** [comment]
At the moment, X-mount.idmap='' is an error (with the above patch to fix errors in general!) but X-mount.idmap=' ' is an empty user and empty group mapping. It feels like they should be the same, one way or the other.

If u: without g: and g: without u: are intentional behaviour to set up a null group/user mapping, I guess that argues that an empty string should be a legitimate way to declare both user and group mappings as null?

**Comment by arachsys** [comment]
Okay, in that case, I guess I should do a quick correctness patch to make the idmap=' ' return the same error as idmap or idmap='', and we can add the proposed extra features like 'null' and 'identity' in a follow-up?

**Comment by brauner** [comment]
> If "null mapping" is possible to achieve and it's supported by the kernel, then we should support it by libmount ;-) , but I'm 

It depends what you mean. It's certainly possible to attach a mapping that maps 0:0, ..., n:n. Is suspect what's happening in the case of the failed idmapping that @arachsys is pointing out though is that no idmapping at all is attached. IOW, a regular bind mount is created.

**Comment by arachsys** [comment]
No, the current behaviour I meant with `-o X-mount.idmap=' '` is a mapping with no users mapped and no groups mapped, i.e. literally a null mapping rather than an identity mapping.

Without the whitespace in the argument, hook_prepare_options() catches this case as the value of the option being null and errors out.

But with the space, at the moment, we set up a namespace and parse zero whitespace-separated tokens to populate its uid_map and gid_map instead.

**Comment by mbwgh** [comment]
I am going to ask this here because I am stuck because of this missing identity mapping.
Apologies if this should go somewhere else.

I want to map `0 --> 1000` and `1000 --> 1000` for both user and group ids.

`man mount` states:
```
The ID-mapping must be specified using the syntax id-type:id-mount:id-host:id-range. Specifying u as the id-type prefix creates a UID-mapping, g creates a GID-mapping and
               omitting id-type or specifying b creates both a UID- and GID-mapping. The id-mount parameter indicates the starting ID in the new mount. The id-host parameter indicates the
               starting ID in the filesystem. The id-range parameter indicates how many IDs are to be mapped. It is possible to specify multiple ID-mappings. The individual ID-mappings
               must be separated by spaces.

               For example, the ID-mapping X-mount.idmap=u:1000:0:1 g:1001:1:2 5000:1000:2 creates an idmapped mount where UID 0 is mapped to UID 1000, GID 1 is mapped to GUID 1001, GID 2
               is mapped to GID 1002, UID and GID 1000 are mapped to 5000, and UID and GID 1001 are mapped to 5001 in the mount.
```

From this, I take that the following should do what I want:
```
$ sudo mount -o ro,X-mount.idmap='u:1000:1000:1 g:1000:1000:1 u:1000:0:1 g:1000:0:1' /dev/mapper/backup-disk /home/mbw/testdir/mnt/
mount: /home/mbw/testdir/mnt/: failed to parse mount options 'ro,X-mount.idmap=u:1000:1000:1 g:1000:1000:1 u:1000:0:1 g:1000:0:1': Invalid argument.
```

I tried some variations (using `b` or omitting the id-type entirely), but always the `Invalid argument` error.
How do I need to format the argument here?

**Comment by brauner** [comment]
You can't map two different IDs to the same ID as that would violate current kernel semantics. We'd need pretty invasive changes to make this possible.
